src/huggingface_hub/_upload_large_folder.py

```diff
@@ -49,6 +49,123 @@
+
+    Warns about:
+        - Too many files in the repository (>100k)
+        - Too many entries (files or subdirectories) in a single folder (>10k)
+        - Files exceeding size limits (>20GB recommended, >50GB hard limit)
+    """
+    logger.info("Running validation checks on files to upload...")
+
+    # Check 1: Total file count
+    if len(paths_list) > MAX_FILES_PER_REPO:
+        logger.warning(
+            f"You are about to upload {len(paths_list):,} files. "
+            f"This exceeds the recommended limit of {MAX_FILES_PER_REPO:,} files per repository.\n"
+            f"Consider:\n"
+            f"  - Splitting your data into multiple repositories\n"
+            f"  - Using fewer, larger files (e.g., parquet files)\n"
+            f"  - See: https://huggingface.co/docs/hub/repositories-recommendations"
+        )
+
+    # Check 2: Files and subdirectories per folder
+    # Track immediate children (files and subdirs) for each folder
+    from collections import defaultdict
+
+    entries_per_folder: Dict[str, Any] = defaultdict(lambda: {"files": 0, "subdirs": set()})
+
+    for paths in paths_list:
+        path = Path(paths.path_in_repo)
+        parts = path.parts
+
+        # Count this file in its immediate parent directory
+        parent = str(path.parent) if str(path.parent) != "." else "."
+        entries_per_folder[parent]["files"] += 1
+
+        # Track immediate subdirectories for each parent folder
+        # Walk through the path components to track parent-child relationships
+        for i in range(len(parts) - 1):  # -1 because last part is the filename
+            if i == 0:
+                # First part is a direct subdirectory of root
+                parent = "."
+                child = parts[0]
+            else:
+                # For nested paths, parent is everything up to index i
+                parent = str(Path(*parts[:i]))
+                child = parts[i]
+
+            # Track this child as a subdirectory of its parent
+            entries_per_folder[parent]["subdirs"].add(child)
+
+    # Check limits for each folder
+    for folder, data in entries_per_folder.items():
+        file_count = data["files"]
+        subdir_count = len(data["subdirs"])
+        total_entries = file_count + subdir_count
+
+        if total_entries > MAX_FILES_PER_FOLDER:
+            folder_display = folder if folder != "." else "root"
+            details = []
+            if file_count > 0:
+                details.append(f"{file_count:,} files")
+            if subdir_count > 0:
+                details.append(f"{subdir_count:,} subdirectories")
+            details_str = " and ".join(details)
+
+            logger.warning(
+                f"Folder '{folder_display}' contains {total_entries:,} entries ({details_str}). "
+                f"This exceeds the recommended limit of {MAX_FILES_PER_FOLDER:,} entries per folder.\n"
+                f"Consider reorganizing into subfolders."
+            )
```

Comment:
(nit) we probably don't need the details temp list here
```suggestion
        if total_entries > MAX_FILES_PER_FOLDER:
            folder_display = "root" if folder == "." else folder
            logger.warning(
                f"Folder '{folder_display}' contains {total_entries:,} entries "
                f"({file_count:,} files and {subdir_count:,} subdirectories). "
                f"This exceeds the recommended {MAX_FILES_PER_FOLDER:,} entries per folder.\n"
                "Consider reorganising into sub-folders."
            )
```